src/Elastic.Apm/OpenTelemetry/ElasticActivityListener.cs

```diff
@@ -0,0 +1,173 @@
+
+				if (KnownListeners.KnownListenersList.Contains(activity.DisplayName))
+					return;
+
+				Transaction transaction = null;
+				if (activity.Context != null && activity.Context.IsRemote && activity.ParentId != null)
+				{
+					var dt = TraceContext.TryExtractTracingData(activity.ParentId.ToString(), activity.Context.TraceState);
+
+					transaction = _tracer.StartTransactionInternal(activity.DisplayName, "unknown",
+						TimeUtils.ToTimestamp(activity.StartTimeUtc), true, activity.SpanId.ToString(),
+						distributedTracingData: dt);
+				}
+				else if (activity.ParentId == null)
+				{
+					transaction = _tracer.StartTransactionInternal(activity.DisplayName, "unknown",
+						TimeUtils.ToTimestamp(activity.StartTimeUtc), true, activity.SpanId.ToString(),
+						activity.TraceId.ToString());
+				}
+				else
+				{
+					Span newSpan;
+					if (_tracer.CurrentSpan == null)
+					{
+						newSpan = (_tracer.CurrentTransaction as Transaction)?.StartSpanInternal(activity.DisplayName, "unknown",
+							timestamp: TimeUtils.ToTimestamp(activity.StartTimeUtc), id: activity.SpanId.ToString());
+					}
+					else
+					{
+						newSpan = (_tracer.CurrentSpan as Span)?.StartSpanInternal(activity.DisplayName, "unknown",
+							timestamp: TimeUtils.ToTimestamp(activity.StartTimeUtc), id: activity.SpanId.ToString());
+					}
+
+					if (newSpan != null)
+					{
+						newSpan.Otel = new OTel();
+						newSpan.Otel.SpanKind = activity.Kind.ToString();
+
+						if (activity.Id != null) ActiveSpans.TryAdd(activity.Id, newSpan);
+					}
+				}
+
+				if (transaction != null)
+				{
+					transaction.Otel = new OTel();
+					transaction.Otel.SpanKind = activity.Kind.ToString();
+
+					if (activity.Kind == ActivityKind.Server)
+						transaction.Type = ApiConstants.TypeRequest;
+					else if (activity.Kind == ActivityKind.Consumer) transaction.Type = ApiConstants.TypeMessaging;
+
+					if (activity.Id != null) ActiveTransactions.TryAdd(activity.Id, transaction);
+				}
+			};
+
+
+		private Action<Activity> ActivityStopped =>
+			activity =>
+			{
+				if (activity == null)
+				{
+					_logger.Trace()?.Log("ActivityStopped called with `null` activity. Ignoring `null` activity.");
+					return;
+				}
+
+				_logger.Trace()?.Log($"ActivityStopped: name:{activity.DisplayName} id:{activity.Id} traceId:{activity.TraceId}");
+
+				if (KnownListeners.KnownListenersList.Contains(activity.DisplayName))
+					return;
+
+				if (activity.Id != null && ActiveTransactions.TryRemove(activity.Id, out var transaction))
+				{
+					transaction.Duration = activity.Duration.TotalMilliseconds;
+
+					if (activity.Tags.Any())
+						transaction.Otel.Attributes = new Dictionary<string, string>();
+
+					foreach (var tag in activity.Tags) transaction.Otel.Attributes.Add(tag.Key, tag.Value);
+					transaction.End();
+				}
+
+				if (activity.Id != null && ActiveSpans.TryRemove(activity.Id, out var span))
```

Comment:
```suggestion
				else if (activity.Id != null && ActiveSpans.TryRemove(activity.Id, out var span))
```

Looks like this could be an `else if` to the previous `if` condition (the branching could also be inside an `if (activity.Id !- null) { /** try remove active transaction, else try remove active span  */ }`

Reply:
Ok - done.